File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/api/ApiServer.java
package systems.zlink.samples.tictactoe.server.api;

import systems.zlink.framework.configuration.ZLinkMeshNodeBuilder;
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode;
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer;
import systems.zlink.samples.tictactoe.server.api.handlers.AuthenticatePlayerHandler;
import systems.zlink.samples.tictactoe.server.configuration.ApiSettings;
import systems.zlink.samples.tictactoe.server.configuration.SampleLogging;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticatePlayerReq;
import systems.zlink.samples.tictactoe.shared.contracts.AuthenticatePlayerRes;

import java.net.URI;

public final class ApiServer {
    private ApiServer() {}

    public static ZLinkFrameworkConfigurer configure(ApiSettings settings) {
        return options -> {
            SampleLogging.configure(settings, "api");
            options.configureDispatch().messageFlow(ZLinkMessageFlowLogMode.NORMAL);
            options.configureLocations();
            URI apiEndpoint = URI.create(settings.apiChannelEndpoint());
            options.addClientServerChannel(SampleNames.ApiChannel)
                    .server()
                    .setBindHost(apiEndpoint.getHost())
                    .listen(apiEndpoint.getPort())
                    // request: AuthenticatePlayerReq에 AuthenticatePlayerRes로 응답한다.
                    .addRequestHandler(
                            AuthenticatePlayerHandler.class,
                            AuthenticatePlayerReq.class,
                            AuthenticatePlayerRes.class);
            ZLinkMeshNodeBuilder mesh = options.addRouteMesh(SampleNames.SpotMesh);
            mesh.listen(settings.routeEndpoint()).setRoutingIdPrefix("tictactoe-api");
            mesh.objects().client();
            // --8<-- [start:doc-manual-peer-connect]
            for (String endpoint : settings.spotEndpoints()) {
                mesh.peerConnections().connect(endpoint);
            }
            // --8<-- [end:doc-manual-peer-connect]
        };
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/api/ApiServer.kt
package systems.zlink.samples.kotlin.tictactoe.server.api

import java.net.URI
import kotlinx.coroutines.Dispatchers
import systems.zlink.contracts.core.RoutingId
import systems.zlink.framework.configuration.ZLinkMessageFlowLogMode
import systems.zlink.framework.kotlin.configureDispatch
import systems.zlink.framework.kotlin.useCoroutineHandlers
import systems.zlink.framework.spring.ZLinkFrameworkConfigurer
import systems.zlink.samples.kotlin.tictactoe.server.api.handlers.AuthenticatePlayerHandler
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleLogging
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleSettings
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerReq
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.AuthenticatePlayerRes

object ApiServer {
    fun configure(settings: SampleSettings): ZLinkFrameworkConfigurer =
        ZLinkFrameworkConfigurer { options ->
            SampleLogging.configure(settings, "api")
            options.useCoroutineHandlers(Dispatchers.Default)
            options.configureDispatch { messageFlow(ZLinkMessageFlowLogMode.NORMAL) }
            options.configureLocations()
            val apiEndpoint = URI.create(settings.apiChannelEndpoint)
            options
                .addClientServerChannel(SampleNames.ApiChannel)
                .server()
                .setBindHost(apiEndpoint.host)
                .listen(apiEndpoint.port)
                // request: AuthenticatePlayerReq에 AuthenticatePlayerRes로 응답한다.
                .addRequestHandler(
                    AuthenticatePlayerHandler::class.java,
                    AuthenticatePlayerReq::class.java,
                    AuthenticatePlayerRes::class.java,
                )
            val mesh = options.addRouteMesh(SampleNames.SpotMesh)
            mesh
                .setRoutingId(RoutingId.from("tictactoe-api-${settings.nodeId}"))
                .listen(settings.routeEndpoint)
            mesh.objects().client()
            // --8<-- [start:doc-manual-peer-connect]
            settings.spotEndpoints.forEachIndexed { index, endpoint ->
                val playNodeId = if (index == 0) "play-a" else "play-b"
                mesh
                    .peerConnections()
                    .connect(RoutingId.from("tictactoe-play-$playNodeId"), endpoint)
            }
            // --8<-- [end:doc-manual-peer-connect]
        }
}
